### PR TITLE
nixos/openvpn3: Add support for systemd-resolved

### PR DESCRIPTION
**Pull request by jfly**
###### Description of changes

I noticed that openvpn3 is been clobbering my `/etc/resolv.conf` file. I dug around a bit, and it turns out that upstream actually does have support for systemd-resolved. I think it makes sense for us to automatically enable that feature if the system is configured to use systemd-resolved.

I opted to not change the default behavior of `pkgs.openvpn3`, but can easily be convinced to change that if folks think I should.

###### Things done

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- [ ] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [x] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [23.11 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2311.section.md) (or backporting [23.05 Release notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2305.section.md))
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

### CODE REVIEW

**Comment by jfly** [comment]
@SCOTT-HAMILTON @KFearsoff any chance either of you could take a look at this? (pinging you because you're listed as openvpn3 maintainers)

**Review by KFearsoff — CHANGES_REQUESTED** [review]
This seems like an overly heavy way to do it, especially since the flag just configures the default value. From what I can tell, openvpn3 supports systemd-resolved out of the box, but it needs to be configured.

From what I can tell, openvpn3 CLI has a config file:
https://github.com/OpenVPN/openvpn3-linux/blob/v20/docs/man/openvpn3-service-netcfg.8.rst.in#L139
We could probably have an option to manage it declaratively.

**Comment by jfly** [comment]
Thanks for the feedback, @KFearsoff! I dug into the configuration file approach, and here's what I've learned:

- I'm not sure yet if we'll still need to add `systemd` to our `buildInputs`. I see a number of build-time checks of `HAVE_SYSTEM` (such as [this code in `openvpn3-admin init-config`](https://github.com/OpenVPN/openvpn3-linux/blob/v20/src/ovpn3cli/commands/init-config.cpp#L269)).
- [The docs](https://github.com/OpenVPN/openvpn3-linux/blob/v20/docs/man/openvpn3-service-netcfg.8.rst.in#L139) on the configuration file(s) are somewhat incomplete (they don't say exactly which files to edit in `@OPENVPN_STATEDIR@`, for instance). I found that running `sudo openvpn3-admin init-config` is a good way to figure out exactly what config files it expects. It even turns out that the `init-config` command is smart enough to [detect when you're running on a system with systemd-resolved](https://github.com/OpenVPN/openvpn3-linux/blob/v20/src/ovpn3cli/commands/init-config.cpp#L269-L289) and to set up a config file accordingly. The [changelog here](https://gitlab.com/openvpn/openvpn3-linux/-/tags/v20) even says that this tool is intended to be used by packaging scripts for various distros, but unfortunately, I don't think it's possible to leverage it on nix.
- As far as I can tell, `OPENVPN_STATEDIR` is a build-time constant. So, it's challenging to change without triggered a rebuild of openvpn3. (If you dig through the code, you'll find various cli flags to set the state dir ([such as these cli params for the `openvpn3-service-netcfg` command](https://github.com/OpenVPN/openvpn3-linux/blob/v20/src/netcfg/openvpn3-service-netcfg.cpp#L362-L366), but we don't control those entrypoints, they're controlled by dbus services files like [`src/service-autostart/net.openvpn.v3.netcfg.service.in`](https://github.com/OpenVPN/openvpn3-linux/blob/v20/src/service-autostart/net.openvpn.v3.netcfg.service.in#L12), which hardcode the `@OPENVPN_STATEDIR@` build-time constant.)

tl;dr: I believe you are right that this can be accomplished with a config file. However, pointing openvpn3 at that config file is a bit trickier.

- We could modify [pkgs/tools/networking/openvpn3/default.nix](https://github.com/NixOS/nixpkgs/blob/master/pkgs/tools/networking/openvpn3/default.nix) to override the `OPENVPN_STATEDIR` build time parameter, but that would force a rebuild of all of openvpn3.
- We could potentially accomplish this with a wrapper derivation where we carefully tweak just the various dbus service files. Searching their repo for [occurrences of `OPENVPN_STATEDIR`](https://github.com/search?q=repo%3AOpenVPN%2Fopenvpn3-linux%20OPENVPN_STATEDIR&type=code) leads me to think that this might actually be possible, it looks like the relevant files are just plaintext: [src/service-autostart/net.openvpn.v3.log.service.in](https://github.com/OpenVPN/openvpn3-linux/blob/90b07a2e816e644c36d1707a9bd4d74244cf210a/src/service-autostart/net.openvpn.v3.log.service.in#L9), [src/service-autostart/net.openvpn.v3.netcfg.service.in](https://github.com/OpenVPN/openvpn3-linux/blob/90b07a2e816e644c36d1707a9bd4d74244cf210a/src/service-autostart/net.openvpn.v3.netcfg.service.in#L9), and [src/service-autostart/net.openvpn.v3.configuration.service.in](https://github.com/OpenVPN/openvpn3-linux/blob/90b07a2e816e644c36d1707a9bd4d74244cf210a/src/service-autostart/net.openvpn.v3.configuration.service.in#L9).
  - That said, this would break if they add more references to `@OPENVPN_STATEDIR@`, and would become impossible if they started embedding it into the c/cpp code.

> This seems like an overly heavy way to do it

Given what I learned from digging into the config file, I suspect the diff currently in this PR (overriding the `DEFAULT_DNS_RESOLVER` build-time flag) is the cleanest way of accomplishing this.

What exactly do you mean by "overly heavy"? Do we want a solution that doesn't involve recompiling openvpn3, even if involves a bit more messiness in terms of the nix code/patches we maintain?

**Comment by KFearsoff** [comment]
> I see a number of build-time checks of `HAVE_SYSTEM` (such as [this code in `openvpn3-admin init-config`](https://github.com/OpenVPN/openvpn3-linux/blob/v20/src/ovpn3cli/commands/init-config.cpp#L269)).

Uhh, this seems bad. Seems like OpenVPN3 wants some thing called `libsystemd-dev` on Ubuntu to be built with resolved support. I *think* we have systemd support already built-in, but I'll have to ask around on Matrix how the stuff that links to systemd is usually built. In the while, do you mind checking if current OpenVPN3 is built with resolved support already?

> The [changelog here](https://gitlab.com/openvpn/openvpn3-linux/-/tags/v20) even says that this tool is intended to be used by packaging scripts for various distros, but unfortunately, I don't think it's possible to leverage it on nix.

I guess we could run it as a `ExecStartPre` on NixOS? Actually, since it's a dbus service, it won't work... But maybe we can do some kind of a one-shot service that would create the config file. Just thinking out load here.

> Given what I learned from digging into the config file, I suspect the diff currently in this PR (overriding the `DEFAULT_DNS_RESOLVER` build-time flag) is the cleanest way of accomplishing this.

Yes, I see how that works now. Your breakdown was perfect! I now also think that's the cleanest way to do it, if we don't go with the oneshot systemd service. I'll investigate if it's a sane way to do it and also see if we actually need to add `systemd` to the package dependencies, and come back with the results!

**Comment by jfly** [comment]
> I guess we could run it as a ExecStartPre on NixOS? Actually, since it's a dbus service, it won't work... But maybe we can do some kind of a one-shot service that would create the config file. Just thinking out load here.

Yeah, that sounds possible, but I'm not sure it's super desirable either, as it wouldn't give us the option of later tweaking any of the defaults that it produces. I imagine that on other OS's, that `openvpn3-admin init-config` is invoked exactly once at package install time, and then humans are free to edit the files it produces after the fact. (I don't know what the story is around future upgrades).

> I'll investigate if it's a sane way to do it and also see if we actually need to add systemd to the package dependencies, and come back with the results!

Thanks! If it's helpful to know, I have been running locally with this patch since July, and it does seem to do what it's supposed to.

**Comment by jfly** [comment]
> I'll investigate if it's a sane way to do it and also see if we actually need to add systemd to the package dependencies, and come back with the results!

@KFearsoff any update here?

**Review by KFearsoff — APPROVED** [review]
Seems like your suggestion is best. Also I tested building without systemd in inputs, and it built, but I'm not sure if it's really working, so let's leave it in inputs.

File: nixos/modules/programs/openvpn3.nix
Comment (Janik-Haag) [inline nixos/modules/programs/openvpn3.nix:16]: are you sure that this is a good idea? because if `config.services.resolved.enable` is true the package is a different one and thus search.nixos.org would lie to you so I would suggest to remove this line
```suggestion
```

**Review by Janik-Haag — COMMENTED** [review]
I would suggest you add a release note entry and then this is probably good to go

File: nixos/modules/programs/openvpn3.nix
Comment (jfly) [reply nixos/modules/programs/openvpn3.nix]: Oh, fair point. Fixed!

**Review by jfly — COMMENTED** [review]
> I would suggest you add a release note entry

@Janik-Haag, done!

File: nixos/modules/programs/openvpn3.nix
Comment (jfly) [reply nixos/modules/programs/openvpn3.nix]: Actually, removing `defaultText` broke building the manual. I've added it back in https://github.com/NixOS/nixpkgs/pull/242371/commits/caa8139199a480cccaadfa0a1639950457901319, but without lying this time!

**Review by Janik-Haag — COMMENTED** [review]
This should be good to go if squash the commits. Into `nixos/openvpn3: Add support for systemd-resolved`

**Comment by Janik-Haag** [comment]
Result of `nixpkgs-review pr 242371` run on x86_64-linux [1](https://github.com/Mic92/nixpkgs-review)
<details>
  <summary>2 packages blacklisted:</summary>
  <ul>
    <li>nixos-install-tools</li>
    <li>tests.nixos-functions.nixos-test</li>
  </ul>
</details>

**Comment by jfly** [comment]
> This should be good to go if squash the commits

@Janik-Haag, squashed!

**Review by Janik-Haag — APPROVED** [review]
LGTM

**Comment by name-snrl** [comment]
> kind of a one-shot service

I did this in my PR, where I also just create a symlink in package to /var. This fixes cli commands perfectly. But that doesn't matter anymore, all I wanted was to fix openvpn on my system with resolved

**Comment by name-snrl** [comment]
@Janik-Haag hey, is there any way to add a package with `enableSystemdResolved = true` to the binary cache? Do we need to add it to all-packages or how? I don't want to build it every time I upgrade

**Comment by Janik-Haag** [comment]
As far as I'm aware the only way would be to split it up into multiple packages because only the default flags get build and pushed by hydra. :(

You can try asking in the `#dev:nixos.org` matrix-room if some one there has another idea to get it in the cache since doing a second package is somewhat meh.

**Pull request merged**
